### PR TITLE
Handle AWS encryption details

### PR DESCRIPTION
**Pull request by artjen**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/artjen/mlflow/pull/12495?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12495/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12495
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Enable MLflow to use AWS server-side encryption details when uploading files.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

For manual tests, I validated the changes in dogfood using an S3 external location that requires KMS parameters. 

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by artjen** [comment]
@mlflow-automation autoformat

**Comment by harupy** [comment]
@mlflow-automation autoformat

**Comment by harupy** [comment]
@mlflow-automation autoformat

File: mlflow/protos/databricks_uc_registry_messages.proto [hunk 1]
```diff
@@ -190,6 +192,26 @@ enum StorageMode {
   DEFAULT_STORAGE = 2;
 }
 
+message EncryptionDetails {
+  oneof encryption_details_type {
+    // Details for CLOUD_MANAGED_SSE_KEYS:
```
Comment (arpitjasa-db) [inline mlflow/protos/databricks_uc_registry_messages.proto:197]: `oneof` just one option? I assume we're trying to match some passed in format exactly, so if this is necessary, let's add a comment explaining why

File: mlflow/protos/databricks_uc_registry_messages.proto [hunk 2]
```diff
@@ -190,6 +192,26 @@ enum StorageMode {
...
+    SseEncryptionDetails sse_encryption_details = 1;
+  }
+}
+
+enum SseEncryptionAlgorithm {
+  SSE_ENCRYPTION_ALGORITHM_UNSPECIFIED = 0;
+  AWS_SSE_KMS = 2;   // "aws:kms" in x-amz-server-side-encryption' header
```
Comment (arpitjasa-db) [inline mlflow/protos/databricks_uc_registry_messages.proto:204]:
```suggestion
  AWS_SSE_KMS = 1;   // "aws:kms" in x-amz-server-side-encryption' header
```
Unless there's a reason why this needs to start at 2?

**Review by arpitjasa-db — COMMENTED** [review]
I know we synced offline about manually testing this, maybe @BenWilson2 knows someone from the field side who has access to an AWS instance for encryption testing. Otherwise, probably file an IT ticket to try this out

File: mlflow/protos/databricks_uc_registry_messages.proto (hunk 2 above)
Comment (artjen) [reply mlflow/protos/databricks_uc_registry_messages.proto:204]: This is to match the UCMR proto, derived from the UC proto. We're only exposing the information we need.

File: mlflow/protos/databricks_uc_registry_messages.proto (hunk 1 above)
Comment (artjen) [reply mlflow/protos/databricks_uc_registry_messages.proto:197]: This is to match the UCMR proto, derived from the UC proto. We're only exposing the information we need.

**Comment by BenWilson2** [comment]
> I know we synced offline about manually testing this, maybe @BenWilson2 knows someone from the field side who has access to an AWS instance for encryption testing. Otherwise, probably file an IT ticket to try this out

@michael-berk would you have access to an AWS account that would make testing this simple? It's a bit challenging for us.

**Comment by artjen** [comment]
@BenWilson2 @michael-berk Any guidance on how to test this manually?

**Comment by artjen** [comment]
@BenWilson2 @michael-berk I was able to validate the changes in dogfood. 

@arpitjasa-db @kriscon-db PTAL.

File: mlflow/store/artifact/optimized_s3_artifact_repo.py [hunk 3]
```diff
@@ -70,6 +72,7 @@ def _refresh_credentials(self):
         self._access_key_id = new_creds["access_key_id"]
         self._secret_access_key = new_creds["secret_access_key"]
         self._session_token = new_creds["session_token"]
+        self._s3_upload_extra_args = new_creds["s3_upload_extra_args"]
```
Comment (kriscon-db) [inline mlflow/store/artifact/optimized_s3_artifact_repo.py:75]: Is this key guaranteed to be there?

File: tests/store/_unity_catalog/model_registry/test_unity_catalog_rest_store.py [hunk 4]
```diff
@@ -314,6 +314,7 @@ def test_create_model_version_with_langchain_dependencies(store, langchain_local
             secret_access_key=secret_access_key,
             session_token=session_token,
             credential_refresh_def=ANY,
+            s3_upload_extra_args={},
```
Comment (kriscon-db) [inline tests/store/_unity_catalog/model_registry/test_unity_catalog_rest_store.py:317]: Any way to add a test/mock around this not being empty every time?

File: mlflow/store/artifact/optimized_s3_artifact_repo.py (hunk 3 above)
Comment (artjen) [reply mlflow/store/artifact/optimized_s3_artifact_repo.py:75]: Yes: the credential refresh function found [here](https://github.com/mlflow/mlflow/pull/12495/files#diff-b8949a6cc9b6ca52729fc888a57fb444676a1f2751737b0a4e54a4b419fa762fR163-R171) ensures this key is present.

File: tests/store/_unity_catalog/model_registry/test_unity_catalog_rest_store.py (hunk 4 above)
Comment (artjen) [reply tests/store/_unity_catalog/model_registry/test_unity_catalog_rest_store.py:321]: Added tests in https://github.com/mlflow/mlflow/pull/12495/commits/d8bd736bd1d02997aa04a436d5067a7aa739022d

**Review by arpitjasa-db — APPROVED** [review]
